packages/zod-form-data/src/helpers.ts

```diff
@@ -10,10 +11,23 @@ import {
   ZodTypeAny,
 } from "zod";
 
+type ExtendsDefaultType<
+  DefaultType extends ZodTypeAny,
+  ProvidedType extends ZodTypeAny
+> = (
+  ProvidedType extends ZodType<any, any, infer Input>
+    ? Input extends input<DefaultType>
+      ? ProvidedType
+      : never
+    : never
+) extends never
+  ? never
+  : ProvidedType;
+
 type InputType<DefaultType extends ZodTypeAny> = {
   (): ZodEffects<DefaultType>;
   <ProvidedType extends ZodTypeAny>(
-    schema: ProvidedType
+    schema: ProvidedType & ExtendsDefaultType<DefaultType, ProvidedType>
```

Comment:
There are two main things to call out here:

### 1.

This might over-limit the schema's we allow. I made a [Typescript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgNwBQFMAnmAKZwCiAHjPQHbEDOAIvZgCGAVwA2MACp16AHgpw4-IWMnS49Npx4oAdAC0SUhgEEONADTy4ABRwA3YMXrEjjDey7ddBl9NM0KAD44AF44AAorWwgHJ18GdU1Pb0NpGUEzczgMizhgDkx6KDgASQ4wYRhAqwUAflLyysSPbWQdfIqYGSURcVdqhUG4eujY51cawYAuOA56OyLJmbmFqAoASmatLxXFutn5vbgZ0cdx6UpqNTLO126BXtUE922U+Pp-YLCEK3D1mbaPiYmEKAGMYNx7so+tJApQFDJTnFXFtkoDUiYzIFIkNuKCABb0ECCE72M7vOAAMmYSR4PRUd3pMIYWSR5wYAzg-zewLBEMRZORsMoaEuoIgHG48HYbBmN0qd3RxAAyjAoPkAOZwAA+bwA8mAYMAJYJRDIlar1RwNYEvogMIIvDlLgB6F1wAACEIAtBoGODfVAcGtxZL4IJQnAZTBwm0OMIQAAjIp-dZUN1wCSE7iMbj4iBiYhwMA4ROCROiGjZUSiOAAd2gAGsKKGpXBE5Ho7GdFKrRrUy2JW3QZ33N3e5q-joIIbjRxTQPW-Ai2Eu21xUVQfQdMRBOxU0A) to help demonstrate.

It covers the main use-cases pretty well, but I think it's reasonable to want something like `z.coerce.date` to work with `z.text`. Unfortunately, `z.coerce` uses the wrong input type on purpose ([reference](https://github.com/colinhacks/zod/issues/1760#issuecomment-1407808567)), so I'm not sure how much we can work around that.

### 2.

It's possible to accomplish this in a simpler way. Since `InputType` is an internal type function (and not exposed as an API), we can freely change the inputs. Here's [another typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgNwBQFMAnmAKZwCSAdmAK4wAqd9APABF6mAIbsANt14A+OAF5EFOHAAUASgBcKAHQAtEgFFMmegGMYAZz7I9JHg0HCxk+-WnTKyvgAUcAN2BiemJXOHoADxh6FmILHX0Q3j4RFhoAGjgU9LghUQkpBncVJWU4C1MAC3oQES1fCACgxIYSzXjDYzNLH39A4NcPCjRKClMIFgt4KMitVg4C-kmoYBYAczgAHzh2GOEV4NkFBAwROKyRgHoLuAABSwBaCIZzR6gcKFHxycz5OGmYFQ2FjsEAAI3oUHUaioVzgXCqFkYFgqEAkxDgYBwoJEoPENEy4nEcAA7tAANafCbwUG-f6A7RLFarKGU76mWkRAE2RlrdTaCBgGDAcYicQssZUuDohR0mxjCGmejaYgiKJQoA) demonstrating an alternate take on this idea. The result appears to be equivalent (unfortunately still having issues with z.coerce).